text/0000-unix-socket.md

```diff
@@ -0,0 +1,491 @@
+    ///
+    /// If the provided value is `None`, then `send` and `send_to` calls will
+    /// block indefinitely. It is an error to pass the zero `Duration` to this
+    /// method.
+    pub fn set_write_timeout(&self, timeout: Option<Duration>) -> io::Result<()> {
+        ...
+    }
+
+    /// Returns the read timeout of this socket.
+    pub fn read_timeout(&self) -> io::Result<Option<Duration>> {
+        ...
+    }
+
+    /// Returns the write timeout of this socket.
+    pub fn write_timeout(&self) -> io::Result<Option<Duration>> {
+        ...
+    }
+
+    /// Moves the socket into or out of nonblocking mode.
+    pub fn set_nonblocking(&self, nonblocking: bool) -> io::Result<()> {
+        ...
+    }
+
+    /// Returns the value of the `SO_ERROR` option.
+    pub fn take_error(&self) -> io::Result<Option<io::Error>> {
+        ...
+    }
+
+    /// Shut down the read, write, or both halves of this connection.
+    ///
+    /// This function will cause all pending and future I/O calls on the
+    /// specified portions to immediately return with an appropriate value
+    /// (see the documentation of `Shutdown`).
+    pub fn shutdown(&self, how: Shutdown) -> io::Result<()> {
+        ...
+    }
+}
+
+impl FromRawFd for UnixDatagram {
+    ...
+}
+
+impl AsRawFd for UnixDatagram {
+    ...
+}
+
+impl IntoRawFd for UnixDatagram {
+    ...
+}
+```
+
+Differences from `UdpSocket`:
+* `bind` takes an `AsRef<Path>` rather than a `ToSocketAddrs`.
+* The `unbound` method creates an unbound socket, as a Unix socket does not need
+  to be bound to send messages.
+* The `pair` method creates a pair of connected, unnamed sockets, as this is
+  commonly used for IPC.
+* The `SocketAddr` returned by the `local_addr` and `peer_addr` methods is
+  different.
+* The `connect`, `send`, `recv`, `set_nonblocking`, and `take_error` methods are
+  not currently present on `UdpSocket` but are provided in the `net2` crate and
+  are being proposed for addition to the standard library in a separate RFC.
+
+## Functionality not present
+
+Some functionality is notably absent from this proposal:
+
+* No support for `SOCK_SEQPACKET` sockets is proposed, as it has not yet been
+  implemented. Since it is connection oriented, there will be a socket type
+  `UnixSeqPacket` and a listener type `UnixSeqListener`. The naming of the
+  listener is a bit unfortunate, but use `SOCK_SEQPACKET` is rare compared to
+  `SOCK_STREAM` so naming priority can go to that version.
+* Unix sockets support file descriptor and credential transfer, but these will
+  not initially be supported as the `sendmsg`/`recvmsg` interface is complex
+  and bindings will need some time to prototype.
+
+These features can bake in the `rust-lang-nursery/unix-socket` as they're
+developed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+While there is precedent for platform specific components in the standard
+library, this will be the by far the largest platform specific addition.
+
```

Comment:
One drawback that may want to be mentioned here is the lack of a smooth migration path from `unix-socket` to the standard library. For example users of Unix sockets will currently have to decide whether to support older versions of Rust (and use `unix-socket`) or newer versions (use `std::os::unix::net`). Additionally, we'll have to duplicate the code when `unix-socket` is compiled against a standard library that also has Unix sockets.

We have ideas for implementing such a migration path, but none of these ideas have an RFC or have been implemented.